File: java/flowcamel-core/src/main/java/com/flowcamel/core/yaml/RouteYamlEmitter.java
package com.flowcamel.core.yaml;

import com.flowcamel.core.model.FlowGraph;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;

public final class RouteYamlEmitter {
  private static final Yaml YAML = new Yaml(dumperOptions());

  private RouteYamlEmitter() {}

  public static String graphToYamlRoutes(FlowGraph graph) {
    List<Map<String, Object>> routes = YamlRouteBuilder.buildAllYamlRoutes(graph);
    if (routes.isEmpty()) return "[]\n";
    return YAML.dump(routes);
  }

  /** Block-style YAML with list items indented under {@code steps} (matches js-yaml / Camel JBang). */
  private static DumperOptions dumperOptions() {
    DumperOptions options = new DumperOptions();
    options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
    options.setIndent(2);
    options.setIndicatorIndent(2);
    options.setIndentWithIndicator(true);
    options.setWidth(120);
    options.setPrettyFlow(true);
    return options;
  }
}

File: packages/core/src/api/RouteYamlEmitter.ts
import yaml from 'js-yaml';
import type { FlowGraph } from '../model/FlowGraph.js';
import { buildAllYamlRoutes } from './YamlRouteBuilder.js';

/** Emit plain Camel YAML DSL (array of routes) for `kamel run *.camel.yaml`. */
export function graphToYamlRoutes(graph: FlowGraph): string {
  const routes = buildAllYamlRoutes(graph);
  if (routes.length === 0) return '[]\n';

  return yaml.dump(routes, {
    lineWidth: 120,
    noRefs: true,
    quotingType: '"',
    sortKeys: (a, b) => {
      if (a === 'steps') return 1;
      if (b === 'steps') return -1;
      return 0;
    },
  });
}
